pg_shard.c

```diff
@@ -530,17 +533,36 @@ ExtractRangeTableEntryWalker(Node *node, List **rangeTableList)
 
 /*
  * DistributedQueryShardList prunes the shards for the table in the query based
- * on the query's restriction qualifiers, and returns this list.
+ * on the query's restriction qualifiers, and returns this list. If the function
+ * cannot find any shards for the distributed table, it errors out. In other sense,
+ * the function errors out or returns a non-empty list.
  */
 static List *
 DistributedQueryShardList(Query *query)
 {
-	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	List *restrictClauseList = NIL;
+	List *prunedShardList = NIL;
 
-	List *restrictClauseList = QueryRestrictList(query);
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
 	List *shardIntervalList = LookupShardIntervalList(distributedTableId);
-	List *prunedShardList = PruneShardList(distributedTableId, restrictClauseList,
-										   shardIntervalList);
+
+	/* error out if no shards exists for the table */
+	if (shardIntervalList == NIL)
+	{
+		char *relName = get_rel_name(distributedTableId);
+
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("could not find any shards for query"),
+						errdetail("No shards exist for distributed table"
```

Comment:
We wrap code at 90 characters, so this full format string can fit on one line here (it's OK that the second argument is orphaned on the next line).